src/Orleans/Streams/Internal/StreamSubscriptionChangeHandler.cs

```diff
@@ -0,0 +1,91 @@
+﻿using Orleans.Runtime;
+using System;
+using Microsoft.Extensions.DependencyInjection;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Orleans.Streams.Core;
+
+namespace Orleans.Streams
+{
+    internal class StreamSubscriptionChangeHandler
+    {
+        private IStreamProviderRuntime providerRuntime;
+        private Dictionary<Type, IStreamSubscriptionObserverProxy> subscriptionObserverMap;
+
+        public StreamSubscriptionChangeHandler(IStreamProviderRuntime providerRT, Dictionary<Type, IStreamSubscriptionObserverProxy> observerMap)
+        {
+            this.providerRuntime = providerRT;
+            this.subscriptionObserverMap = observerMap;
+        }
+
+        /// <summary>
+        /// This method finds the correct IStreamSubscriptionObserver<T> based on item's type, and invoke its OnNewSubscription method
+        /// </summary>
+        /// <param name="subscriptionId"></param>
+        /// <param name="streamId"></param>
+        /// <param name="messageType"></param>
+        /// <returns></returns>
+        public async Task HandleNewSubscription(GuidId subscriptionId, StreamId streamId, Type messageType)
+        {
+            // if no observer attached to the subscription, check if the grain is a IStreamSubscriptionObserver
+            IStreamSubscriptionObserverProxy observerProxy;
+            if (TryGetStreamSubscriptionObserverProxyForType(messageType, out observerProxy))
+            {
+                var streamProvider = this.providerRuntime.ServiceProvider
```

Comment:
doesn't seem like we need providerRuntime  or the service provider here, we only need the stream provider manager, so why not take only that in the constructor?